### PR TITLE
Add baseline stubs for Python Tensor types

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Base for the Python Tensor and TensorList types interface files,
generated using stubgen. 

C-based types (or types not guaranteed to be present, like np.array), 
are removed from signatures and left as a comment or replaced with Any.

This doesn't necessary add new information, just enables static tools
to access the types. 

```
    imgs, labels = pipe.run()
    reveal_type(imgs)
```

Before:
```
type_annotations/test_typing_pipelines.py:63: note: Revealed type is "Union[Any, Any]"
```

After:
```
type_annotations/test_typing_pipelines.py:63: note: Revealed type is "Union[nvidia.dali.backend_impl.TensorListCPU, nvidia.dali.backend_impl.TensorListGPU]"
```

Note that the backend contains many more types, like OpSchema, OpSpec, Pipeline, etc.
They are internal, so I didn't expose them, but we should consider adding them to aid our development.

This file is generated and than manually adjusted, I don't see a clear automated workflow for this.
It appears harder to maintain than a list of supported enum values.

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/python/nvidia/dali/backend_impl.pyi [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    @overload
+    def __init__(self, b, layout: str = ..., is_pinned: bool = ...) -> None:
+        ...
+
+    def _as_gpu(self, *args, **kwargs) -> Any:
+        ...
+
+    # def _expose_dlpack_capsule(self) -> capsule: ...
+    def _expose_dlpack_capsule(self) -> Any:
+        ...
```
Comment (mzient) [inline dali/python/nvidia/dali/backend_impl.pyi:39]: It starts with underscore, so it's apparently private. Do we want to document that?

File: dali/python/nvidia/dali/backend_impl.pyi [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+    # @overload
+    # def __init__(self, b: buffer, layout: str = ..., is_pinned: bool = ...) -> None: ...
+    @overload
+    def __init__(self, b, layout: str = ..., is_pinned: bool = ...) -> None:
+        ...
+
+    def _as_gpu(self, *args, **kwargs) -> Any:
+        ...
```
Comment (mzient) [inline dali/python/nvidia/dali/backend_impl.pyi:35]: Do we want to document private members, especially ones as generic as this one?

File: dali/python/nvidia/dali/backend_impl.pyi [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    @overload
+    def __init__(self, b, layout: str = ..., is_pinned: bool = ...) -> None:
+        ...
+
+    @overload
+    def __init__(self, list_of_tensors: list, layout: str = ...) -> None:
+        ...
+
+    def _as_gpu(self, *args, **kwargs) -> Any:
+        ...
```
Comment (mzient) [inline dali/python/nvidia/dali/backend_impl.pyi:147]: Likewise.

File: dali/python/nvidia/dali/backend_impl.pyi (hunk 1 above)
Comment (klecki) [reply dali/python/nvidia/dali/backend_impl.pyi:39]: Good point, I planned to remove the private ones from here, and forgot to do that.

**Review by banasraf — APPROVED** [review]

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/backend_impl.pyi (hunk 2 above)
Comment (klecki) [reply dali/python/nvidia/dali/backend_impl.pyi]: done

**Comment by klecki** [comment]
After the last changes, mypy declares the return types of `Pipeline.run()` as 
> Revealed type is "builtins.tuple[Union[nvidia.dali.tensors.TensorListCPU, nvidia.dali.tensors.TensorListGPU], ...]"

I am intentionally not adjusting any other imports to see if this change doesn't break anything in the existing code. 
It would be good to check the generated docs as well.

**Comment by klecki** [comment]
>  It would be good to check the generated docs as well.
the docs set the module explicitly, but still, even though we adjusted the module path, the signatures utilize the backend_impl variant.

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
obraz
It works

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/tensors.pyi [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    # def __init__(self, b: buffer, layout: str = ..., is_pinned: bool = ...) -> None: ...
+    @overload
+    def __init__(self, b, layout: str = ..., is_pinned: bool = ...) -> None:
+        ...
+
+    def _as_gpu(self, *args, **kwargs) -> Any:
+        ...
+
+    # def _expose_dlpack_capsule(self) -> capsule: ...
+        ...
```
Comment (stiepan) [inline dali/python/nvidia/dali/tensors.pyi:37]: ...

Comment (stiepan) [reply dali/python/nvidia/dali/tensors.pyi:37]: shouldn't this get commented out as well?

**Review by stiepan — APPROVED** [review]
FYI: I did not look into the documentation as it's not built yet.

File: dali/python/nvidia/dali/tensors.pyi (hunk 4 above)
Comment (klecki) [reply dali/python/nvidia/dali/tensors.pyi:37]: It probably should :)

**Comment by klecki** [comment]
!build

**Pull request merged**